src/transport/TCP.cpp

```diff
@@ -0,0 +1,617 @@
+    endPoint->OnConnectionClosed   = OnConnectionClosed;
+    endPoint->OnConnectionReceived = OnConnectionReceived;
+    endPoint->OnAcceptError        = OnAcceptError;
+    endPoint->OnPeerClose          = OnPeerClosed;
+
+    err = endPoint->Connect(addr.GetIPAddress(), addr.GetPort(), addr.GetInterface());
+    SuccessOrExit(err);
+
+    // enqueue the packet once the connection succeeds
+    packet->peerAddress  = addr;
+    packet->packetBuffer = msg;
+    msg                  = nullptr;
+    mUsedEndPointCount++;
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        if (msg != nullptr)
+        {
+            System::PacketBuffer::Free(msg);
+            msg = nullptr;
+        }
+        if (endPoint != nullptr)
+        {
+            endPoint->Free();
+        }
+    }
+    return err;
+}
+
+CHIP_ERROR TCPBase::ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, System::PacketBuffer * buffer,
+                                                       uint16_t messageSize)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    uint8_t * oldStart = buffer->Start();
+    size_t oldLength   = buffer->DataLength();
+
+    buffer->SetDataLength(messageSize);
+
+    size_t headerSize = 0;
+
+    MessageHeader header;
+    err = header.Decode(buffer->Start(), buffer->DataLength(), &headerSize);
+    SuccessOrExit(err);
+
+    buffer->ConsumeHead(headerSize);
+
+    // message receive handler will attempt to free the buffer, however as the buffer may
+    // contain additional data, we retain it to prevent actual free
+    buffer->AddRef();
+
+    HandleMessageReceived(header, peerAddress, buffer);
+
+exit:
+    buffer->SetStart(oldStart);
+    buffer->SetDataLength(oldLength);
+
+    return err;
+}
+
+CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress,
+                                          System::PacketBuffer * buffer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    while (buffer != nullptr)
+    {
+        // when a buffer is empty, it can be released back to the app
+        if (buffer->DataLength() == 0)
+        {
+            System::PacketBuffer * old = buffer;
+            buffer                     = old->DetachTail();
+            System::PacketBuffer::Free(old);
+            continue;
+        }
+
+        uint8_t * messageData = nullptr;
+        uint16_t messageSize  = 0;
+        if (ContainsCompleteMessage(buffer, &messageData, &messageSize))
+        {
+            // length was read and is not needed anymore
+            buffer->ConsumeHead(kPacketSizeBytes);
+
+            // Sanity checks. These are more like an assert for invariants
+            VerifyOrExit(messageData == buffer->Start(), err = CHIP_ERROR_INTERNAL);
+            VerifyOrExit(buffer->DataLength() >= messageSize, err = CHIP_ERROR_INTERNAL);
+
+            err = ProcessSingleMessageFromBufferHead(peerAddress, buffer, messageSize);
+            buffer->ConsumeHead(messageSize);
```

Comment:
I thought to keep it because I alrady skipped over the size. If processsinglemessage fails on a buffer, there seems to be reasonable expectation that it will fail in the future as well. This also guarantees that we empty buffers in time.

I.e. I think this is correct and it would be more work to not consume on error.

Reply:
OK.  That seems fine, but worth a comment explaining why we are doing what we are doing.